### ISSUE TITLE
Wallet sync gets stuck

### ISSUE THREAD

**Issue by bonomat** [itchysats/itchysats#2287]
It looks like the wallet sync gets stuck after the first successful sync: 

```
{"timestamp":"2022-06-23T07:31:24.499078676Z","level":"TRACE","fields":{"message":"Wallet sync started"},"target":"wallet"}

...
{"timestamp":"2022-06-23T07:31:48.555812439Z","level":"DEBUG","fields":{"message":"Saving `sync_time` = SyncTime { block_time: BlockTime { height: 2281901, timestamp: 1655969508 } }","log.target":"bdk::wallet","log.module_path":"bdk::wallet","log.file":"/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bdk-0.18.0/src/wallet/mod.rs","log.line":1594},"target":"bdk::wallet"}
{"timestamp":"2022-06-23T07:31:48.611153317Z","level":"TRACE","fields":{"message":"Wallet sync done","sync_time_sec":"24"},"target":"wallet"}

...
{"timestamp":"2022-06-23T07:34:48.612697524Z","level":"TRACE","fields":{"message":"Wallet sync started"},"target":"wallet"}
{"timestamp":"2022-06-23T07:34:48.612747376Z","level":"DEBUG","fields":{"message":"Begin sync...","log.target":"bdk::wallet","log.module_path":"bdk::wallet","log.file":"/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bdk-0.18.0/src/wallet/mod.rs","log.line":1570},"target":"bdk::wallet"}
{"timestamp":"2022-06-23T07:34:48.612768043Z","level":"DEBUG","fields":{"message":"max_address 100","log.target":"bdk::wallet","log.module_path":"bdk::wallet","log.file":"/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bdk-0.18.0/src/wallet/mod.rs","log.line":360},"target":"bdk::wallet"}
{"timestamp":"2022-06-23T07:34:48.613134372Z","level":"DEBUG","fields":{"message":"run_setup: false","log.target":"bdk::wallet","log.module_path":"bdk::wallet","log.file":"/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bdk-0.18.0/src/wallet/mod.rs","log.line":1577},"target":"bdk::wallet"}
```

It looks like it just gets stuck somewhere inside the wallet and does not go on.

**Comment by bonomat** [comment]
I have two ideas: 

1. it could be the electrum "fix" added in https://github.com/itchysats/itchysats/commit/c7c285520567dc69c08a562a7a887f80125fca4e
2. it has something to do with the database

I assume it's the latter one because of one symptom we found locally when adding db support to the wallet: 

Initially we wanted to go with an SQL db but had some weird errors: `database is locked` or similar. We thought it's an issue with sqlite and hence went with sled. 
What if the problem exists as well with sled and the wallet is in a locked state and we end up in a deadlock? 

The weird thing is that I could not reproduce this case locally on my machine. Syncing seems to just work. 

Re the SQL problem of `database is locked`: I created a standalone PoC which never ran into this problem. Only our daemon runs into this problem.

**Comment by klochowicz** [comment]
> I have two ideas:
> 
> 1. it could be the electrum "fix" added in [c7c2855](https://github.com/itchysats/itchysats/commit/c7c285520567dc69c08a562a7a887f80125fca4e)
> 2. it has something to do with the database
> 
> I assume it's the latter one because of one symptom we found locally when adding db support to the wallet:
> 
> Initially we wanted to go with an SQL db but had some weird errors: `database is locked` or similar. We thought it's an issue with sqlite and hence went with sled. What if the problem exists as well with sled and the wallet is in a locked state and we end up in a deadlock?
> 
> The weird thing is that I could not reproduce this case locally on my machine. Syncing seems to just work.
> 
> Re the SQL problem of `database is locked`: I created a standalone PoC which never ran into this problem. Only our daemon runs into this problem.

a tricky bug indeed. Perhaps we could revert the wallet DB and see whether the problem is still there? (then at least we would have more data about the problem?)

**Comment by bonomat** [comment]
For the cost of it, I would try 1. first and see if it helps.

**Comment by klochowicz** [comment]
> For the cost of it, I would try 1. first and see if it helps.

Sure, let's give it a shot - seems like no-one reported similar issue yet, despite this fix being released upstream 2 weeks ago.  at least in the process of trying we can remove the fork :)

**Comment by klochowicz** [comment]
I have actioned the idea 1) and it seems to have fixed the problem (at least for now): #2289 

Closing for now, will reopen if it happens again after letting it run for a bit.
